src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java

```diff
@@ -1209,8 +1226,19 @@ else if (!referencePath.isDirectory())
                     String sModulePath = getSubmodulePath(sModuleName);
 
                     perModuleArgs.add(sModulePath);
-                    launchCommandWithCredentials(perModuleArgs, workspace, cred, urIish, timeout);
+                    StandardCredentials finalCred = cred;
+                    URIish finalUrIish = urIish;
+                    executorService.submit(() -> {
+                        try {
+                            launchCommandWithCredentials(perModuleArgs, workspace, finalCred, finalUrIish, timeout);
+                        } catch (InterruptedException e) {
+                            throw new GitException("Interrupted while updating submodule for " + sModuleName);
+                        }
+                    });
                 }
+
+                executorService.shutdown();
+                executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
```

Comment:
There are more flaws in here:

1) The timeout value is really big. Why not use the configurable `timeout` value?
2) The method [boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#awaitTermination-long-java.util.concurrent.TimeUnit-) just asks the tasks to complete. If they have not been completed within the specified timeout this is returned in the result. [List<Runnable> shutdownNow()](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#shutdownNow--) should then be used.
3) `InterruptedException`s are not handled.

A good example to properly use the `ExectorService` is given in its [Javadocs](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html) in the beginning.

Reply:
1. As I understand it (based on the previous implementation), the `timeout` value is for each submodule update, not the duration of all updates. If we used `timeout` for the value in `awaitTermination`, it would then apply to the duration of all submodules.
2. Do you mean to say that we should check the return value of `shutdown()` and then call `shutdownNow()` if it returned false?
3. I didn't think that this was necessary because the method declares that it throws `InterruptedException`. Should we rather `try`/`catch` the `awaitTermination` statement and repackage the `InterruptedException` as a `GitException`?